filebeat/input/file/state.go

```diff
@@ -81,25 +83,36 @@ func (s *States) findPrevious(newState State) (int, State) {
 }
 
 // Cleanup cleans up the state array. All states which are older then `older` are removed
-func (s *States) Cleanup(older time.Duration) {
+func (s *States) Cleanup() {
+
 	s.mutex.Lock()
 	defer s.mutex.Unlock()
 
 	for i, state := range s.states {
 
-		// File wasn't seen for longer then older -> remove state
-		if time.Since(state.LastSeen) > older {
-			logp.Debug("prospector", "State removed for %s because of older: %s", state.Source)
-			s.states = append(s.states[:i], s.states[i+1:]...)
+		// -1 means ttl is disabled
+		if state.Ttl == -1 {
+			continue
 		}
-	}
 
+		// States is older then ttl
+		if int64(time.Since(state.Timestamp).Seconds()) > state.Ttl {
+			// This check prevents a panic
+			if len(s.states) == 1 {
+				s.states = []State{}
+			} else {
+				s.states = append(s.states[:i], s.states[i+1:]...)
+			}
```

Comment:
filter loop alternative (not as many potential copies on sub-splices when removing multiple elements):

```
states := s.states[:0]
for _, state := range.stats {
    ttl := state.Ttl * time.Second
    if ttl >= 0 && time.Since(state.Timestamp) <= ttl {
          debugf("State removed for %v because of older: %v", state.Source, ttl)
          continue // drop state
    }
    states = append(states, state) // in-place copy old state
}
s.states = states
```

When converting ttl into time.Duration it nicely prints the unit being used. time.Duration can be negative too.